File: submission/hrishikesh/src/test/java/com/kulu/sdet/support/DbAssertions.java
package com.kulu.sdet.support;

import java.util.List;
import java.util.Map;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;

/**
 * Reusable, intent-revealing database assertions. Every read hits the real DB so tests prove that
 * API-visible state matches persisted state.
 */
@Component
public class DbAssertions {

  private final JdbcTemplate jdbc;

  public DbAssertions(JdbcTemplate jdbc) {
    this.jdbc = jdbc;
  }

  public long balanceOf(String walletId) {
    Long b = jdbc.queryForObject("SELECT balance FROM wallets WHERE id = ?", Long.class, walletId);
    if (b == null) {
      throw new AssertionError("wallet not found: " + walletId);
    }
    return b;
  }

  public int transferCount() {
    Integer n = jdbc.queryForObject("SELECT COUNT(*) FROM transfers", Integer.class);
    return n == null ? 0 : n;
  }

  public int transferCountForIdempotencyKey(String key) {
    Integer n =
        jdbc.queryForObject(
            "SELECT COUNT(*) FROM transfers WHERE idempotency_key = ?", Integer.class, key);
    return n == null ? 0 : n;
  }

  public Map<String, Object> transferRow(String transferId) {
    return jdbc.queryForMap("SELECT * FROM transfers WHERE id = ?", transferId);
  }

  public int idempotencyRowCount(String key) {
    Integer n =
        jdbc.queryForObject(
            "SELECT COUNT(*) FROM idempotency_keys WHERE idempotency_key = ?", Integer.class, key);
    return n == null ? 0 : n;
  }

  public Map<String, Object> idempotencyRow(String key) {
    return jdbc.queryForMap(
        "SELECT idempotency_key, payload_hash, transfer_id, response_status, response_body"
            + " FROM idempotency_keys WHERE idempotency_key = ?",
        key);
  }

  public int auditCount(String transferId, String eventType) {
    Integer n =
        jdbc.queryForObject(
            "SELECT COUNT(*) FROM transfer_events WHERE transfer_id = ? AND event_type = ?",
            Integer.class,
            transferId,
            eventType);
    return n == null ? 0 : n;
  }

  public List<Map<String, Object>> auditRows(String transferId) {
    return jdbc.queryForList(
        "SELECT id, transfer_id, event_type, payload, created_at FROM transfer_events"
            + " WHERE transfer_id = ? ORDER BY id",
        transferId);
  }

  public int outboxCount(String transferId) {
    Integer n =
        jdbc.queryForObject(
            "SELECT COUNT(*) FROM outbox_events WHERE aggregate_id = ?", Integer.class, transferId);
    return n == null ? 0 : n;
  }

  public int outboxCount(String transferId, String eventType) {
    Integer n =
        jdbc.queryForObject(
            "SELECT COUNT(*) FROM outbox_events WHERE aggregate_id = ? AND event_type = ?",
            Integer.class,
            transferId,
            eventType);
    return n == null ? 0 : n;
  }

  public List<Map<String, Object>> outboxRows(String transferId) {
    return jdbc.queryForList(
        "SELECT id, aggregate_id, event_type, payload, published_at FROM outbox_events"
            + " WHERE aggregate_id = ? ORDER BY id",
        transferId);
  }
}

File: submission/neha/tests/helpers/db_assertions.py
"""Reusable database assertions for transfer invariants."""

from __future__ import annotations


def wallet_balance(db, wallet_id: str) -> int:
    row = db.execute(
        "SELECT balance FROM wallets WHERE id = ?",
        (wallet_id,),
    ).fetchone()
    assert row is not None, f"wallet {wallet_id} missing"
    return row["balance"]


def transfer_count(db) -> int:
    return db.execute("SELECT COUNT(*) AS c FROM transfers").fetchone()["c"]


def audit_count(db, transfer_id: str | None = None) -> int:
    if transfer_id is None:
        return db.execute("SELECT COUNT(*) AS c FROM audit_events").fetchone()["c"]
    return db.execute(
        "SELECT COUNT(*) AS c FROM audit_events WHERE transfer_id = ?",
        (transfer_id,),
    ).fetchone()["c"]


def outbox_count(db, transfer_id: str | None = None) -> int:
    if transfer_id is None:
        return db.execute("SELECT COUNT(*) AS c FROM outbox_events").fetchone()["c"]
    return db.execute(
        "SELECT COUNT(*) AS c FROM outbox_events WHERE transfer_id = ?",
        (transfer_id,),
    ).fetchone()["c"]


def idempotency_row(db, key: str):
    return db.execute(
        "SELECT * FROM idempotency_keys WHERE key = ?",
        (key,),
    ).fetchone()


def assert_balances(
    db, *, source: str, destination: str, source_bal: int, dest_bal: int
):
    assert wallet_balance(db, source) == source_bal
    assert wallet_balance(db, destination) == dest_bal


def assert_no_side_effects(db):
    assert transfer_count(db) == 0
    assert audit_count(db) == 0
    assert outbox_count(db) == 0
    count = db.execute("SELECT COUNT(*) AS c FROM idempotency_keys").fetchone()["c"]
    assert count == 0


def assert_successful_transfer_persistence(db, transfer_id: str, *, amount: int):
    transfer = db.execute(
        "SELECT * FROM transfers WHERE id = ?",
        (transfer_id,),
    ).fetchone()
    assert transfer is not None
    assert transfer["status"] == "completed"
    assert transfer["amount"] == amount

    assert audit_count(db, transfer_id) == 1
    audit = db.execute(
        "SELECT * FROM audit_events WHERE transfer_id = ?",
        (transfer_id,),
    ).fetchone()
    assert audit["event_type"] == "transfer_completed"

    assert outbox_count(db, transfer_id) == 1
    outbox = db.execute(
        "SELECT * FROM outbox_events WHERE transfer_id = ?",
        (transfer_id,),
    ).fetchone()
    assert outbox["event_type"] == "wallet.transfer.completed"
    assert outbox["status"] == "pending"
